Lazy imports up in here

usergenic opened the pull request:
Rewrote `buildDepsIndex` to do-the-right-thing™ wrt `lazy-import`.

- BREAKING: Bundler now supports "lazy imports" aka `<link rel="lazy-import">`.
  Imports which are lazy-imported are now implicitly treated as entrypoints.
- [x] CHANGELOG.md has been updated

usergenic commented:
Fixes https://github.com/Polymer/polymer-bundler/issues/447

rictic commented on src/deps-index.ts:
```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
-      return {url: url, eager: eagerImports, lazy: lazyImports};
+  // Gets a collection of only the direct html imports for the document.
+  const htmlImports = document.getFeatures(
+      {kind: 'html-import', imported: false, externalPackages: true});
```
hm, I think this should be named `getTransitiveHtmlDependencies` then

FredKSchott commented on src/deps-index.ts:
```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
+    analysis: Analysis,
+    visited: Set<UrlString> = new Set<UrlString>()): DependencyMapEntry {
+  const map: DependencyMapEntry = {
+    eager: new Set(),
```
helpful to add a type here. Set() of what?

usergenic replied:
Done.

FredKSchott commented on src/deps-index.ts:
```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
+  const unprocessed = new Set<UrlString>(entrypoints);
+  unprocessed.forEach((entrypoint) => {
+    processed.add(entrypoint);
+    unprocessed.delete(entrypoint);
```
oof, editting the object being iterated over always gives me the creeps. Iterate over `entrypoints` or `new Set(entrypoints)` here instead

usergenic replied:
Well, I'm iterating over the set I'm mutating because I want to add things to it like a queue as in https://github.com/Polymer/polymer-bundler/pull/487#discussion-diff-113536810R116

FredKSchott commented on src/deps-index.ts:
```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
+
+    // Add lazy edges as unprocessed entrypoints, if they haven't already been
+    // processed.
+    deps.lazyEdges.forEach((dep) => processed.has(dep) || unprocessed.add(dep));
```
can you expand this to be more clear? new lines are free :)

usergenic replied:
Now that I don't use processed and unprocessed, this is simpler:
```typescript
    // Add lazy imports as entrypoints to index.
    for (const dep of deps.lazyImports) {
      allEntrypoints.add(dep);
    }
  ```

FredKSchott commented on src/import-utils.ts:
```diff
@@ -93,8 +105,13 @@ export async function inlineHtmlImport(
 
-  // If the analyzer could not load the import document, we can't inline it, so
-  // lets skip it.
+  // We will stop short of actually inlining a `lazy-import` or its children.
```
Why? Would be useful to know for future devs

usergenic replied:
Updated comment:

```typescript
  // We don't actually inline a `lazy-import` because its loading is intended
  // to be deferred until the client requests it.
```

rictic commented on src/deps-index.ts:
```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
+        (dep) => map.eager.has(dep) ? map.lazy.delete(dep) : map.lazy.add(dep));
+    deps.lazy.forEach((dep) => map.eager.has(dep) || map.lazy.add(dep));
+    deps.lazyEdges.forEach((dep) => map.lazyEdges.add(dep));
+  });
```
This algorithm isn't obviously correct to me. Here's my thinking based on what I've inferred from what we're trying to do here, lemmy know if I go off track:

A file has eager and lazy imports of other files. A file is an eager dependency of another if that file is reachable by any path of entirely eager imports. A file is a lazy dependency of another if all paths to that file include at least one lazy import.

This means that every eager dependency of an eager dependency is an eager dependency, and the rest of its dependencies are lazy. So I think an easier to understand implementation would involve building the set of files reachable by by entirely eager paths, and another set of all files reachable at all, then constructing the lazy files as those that are in the set of all files but not in the set of eager files.

I think this algorithm is equivalent in complexity but simpler to reason about.

justinfagnani replied:
Some light afternoon reading: https://www.kent.edu/sites/default/files/TR-KSU-CS-2010-01.pdf

We should be able to do this in a single BFS, with a change that a node can be visited more than once if we're seeing it via an eager edge for the first time, essentially performing two reachability queries at the same time:

Something like this modified BFS should work:

```js
getHtmlDependencies(document) {
  const deps = new Set();
  const eagerDeps = new Set();
  const lazyImports = new Set();
  _getHtmlDependencies(document, true, deps, eagerDeps, lazyImports);
  return {deps, eagerDeps, lazyImports};
}

_getHtmlDependencies(document, viaEager, visited, visitedEager, lazyImports) {
  const imports = document.getFeatures({kind: 'html-import', imported: false, externalPackages: true});
  for (const import of imports) {
    const importUrl = import.document.url;
    if (import.lazy) {
      lazyImports.add(importUrl);
    }
    if (visitedEager.has(importUrl)) {
      continue;
    }
    const isEager = viaEager || !import.lazy;
    if (isEager) {
      visitedEager.add(importUrl);
      // In this case we've visited a node eagerly for the first time,
      // so recurse
    } else if (visited.has(importUrl)) {
      // In this case we're seeing a node lazily again, so don't recurse
      continue;
    }
    visited.add(importUrl);
    getHtmlDependencies(import.document, isEager, visited, visitedEager, lazyImports);
  }
}
```

usergenic replied:
I totally used this.

justinfagnani commented on CHANGELOG.md:
```diff
@@ -6,6 +6,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/)
 and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## Unreleased
+- BREAKING: Bundler now supports "lazy imports" aka `<link rel="lazy-import">`.
```
Breaking or new feature?

usergenic replied:
Updated CHANGELOG comment to explain why it was BREAKING.

justinfagnani commented on src/deps-index.ts:
```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
+    lazyEdges: new Set()
+  };
+  const document = analysis.getDocument(url);
+  if (!(document instanceof Document)) {
```
we do this all over right now, because getDocument() may return a warning

usergenic replied:
I *kinda* want an `analysis.getDocumentOrBust(url);` option to dry up these redundant go-like error-handling clauses.

rictic commented on src/deps-index.ts:
```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
+  const analysis = await analyzer.analyze(entrypoints);
+  const processed = new Set<UrlString>();
+  const unprocessed = new Set<UrlString>(entrypoints);
+  unprocessed.forEach((entrypoint) => {
```
prefer `for ... of` over `.forEach`

usergenic replied:
i'm replacing all of them.

usergenic replied:
replaced all of them.

rictic commented on src/deps-index.ts:
```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
+    unprocessed.delete(entrypoint);
+    const deps = getTransitiveDependencies(entrypoint, analysis);
+    depsIndex.entrypointToDeps.set(
+        entrypoint, new Set([entrypoint].concat(Array.from(deps.eager))));
```
golf: `[entrypoint].concat(Array.from(deps.eager))` -> `[entrypoint, ...deps.eager]`

usergenic replied:
+1

usergenic replied:
LIKE.  Using.

rictic commented on src/deps-index.ts:
```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
+    visited: Set<UrlString> = new Set<UrlString>()): DependencyMapEntry {
+  const map: DependencyMapEntry = {
+    eager: new Set(),
+    lazy: new Set(),
```
is `lazy` ever used? or is that for a future PR?

usergenic replied:
this is gone.  replaced with new type.

rictic commented on src/deps-index.ts:
```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
+  const depsIndex = {entrypointToDeps: new Map<UrlString, Set<UrlString>>()};
+  const analysis = await analyzer.analyze(entrypoints);
+  const processed = new Set<UrlString>();
+  const unprocessed = new Set<UrlString>(entrypoints);
```
Because it's a set, if you only added entries to `unprocessed` and never deleted then you could get rid of `processed` entirely. Could rename it from `unprocessed` to just `urls`

usergenic replied:
unprocessed is really a queue.  it's added to within the loop that iterates over it.  i'll look at rewriting to make that more obvious.

usergenic replied:
(its a queue that doesn't care about order.)

rictic replied:
I got that, but you don't need `processed`, and you don't need to delete from `unprocessed`

usergenic replied:
Done.  Renamed to `allEntrypoints`

usergenic commented on src/deps-index.ts:
```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
     }
+  });
+
+  visited.add(url);
```
moot, because replaced with your awsm code

usergenic commented on src/deps-index.ts:
```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
+  Array.from(map.eager).filter((d) => !visited.has(d)).forEach((d) => {
+    const deps = getTransitiveDependencies(d, analysis, visited);
+    deps.eager.forEach((dep) => map.eager.add(dep) && map.lazy.delete(dep));
+    deps.lazy.forEach((dep) => map.eager.has(dep) || map.lazy.add(dep));
```
this has been deleted in favor of new shorter algorithm.

usergenic commented on src/bundler.ts:
```diff
@@ -227,10 +226,8 @@ export class Bundler {
   private async _bundleDocument(
       docBundle: AssignedBundle,
-      bundleManifest: BundleManifest,
-      bundleImports?: Set<string>): Promise<ASTNode> {
```
the tslint file has `        "no-unused-variable": true,`

rictic replied:
tsconfig is the place to do this

usergenic replied:
Ok.  Adding in separate PR so it doesn't muddy this one. (there are a few other spots including exported functions with unused parameters.)

usergenic replied:
https://github.com/Polymer/polymer-bundler/issues/490

justinfagnani commented on src/deps-index.ts:
```diff
@@ -21,79 +20,86 @@ export interface DepsIndex {
+  const analysis = await analyzer.analyze(entrypoints);
+  const allEntrypoints = new Set<UrlString>(entrypoints);
+  for (const entrypoint of allEntrypoints) {
+    const document = analysis.getDocument(entrypoint);
```
maybe name `documentOrWarning`

usergenic replied:
Done.

justinfagnani commented on src/deps-index.ts:
```diff
@@ -21,79 +20,86 @@ export interface DepsIndex {
+    const document = analysis.getDocument(entrypoint);
+    if (!(document instanceof Document)) {
+      const message = document && document.message || '';
+      throw new Error(`Unable to get document ${entrypoint}: ${message}`);
```
I don't think we should throw in this case. Ideally we only throw for program errors, and bad user input is handled in normal execution and warnings and errors are threaded through the results. Here we should continue building if possible and just warn and exit with a non-0 code. It would be a bit of work to do that now, probably changing the return type and such, so maybe we can just `console.warn` for now.

usergenic replied:
Ok.  Just console.warn'n for now.

usergenic commented:
@justinfagnani @rictic @FredKSchott Addressed all PR feedback.  Also, gutted and replaced the `getTransitiveDependencies` method with the more efficient one from Justin.

justinfagnani commented on src/deps-index.ts:
```diff
@@ -21,79 +20,86 @@ export interface DepsIndex {
+
+    // Add lazy imports as entrypoints to index.
+    for (const dep of deps.lazyImports) {
+      allEntrypoints.add(dep);
```
does this cause the for loop to iterate over the lazy import? Do we need to process it? I'd consider making this explicit by making a queue of entrypoints and adding the lay import to the queue.

usergenic replied:
I added a comment to call it out, since it was surprising (and awesome).

justinfagnani commented on src/import-utils.ts:
```diff
@@ -77,11 +79,22 @@ export async function inlineHtmlImport(
+        // means we've handled that entire bundle, so we guard against inlining
+        // any other file from that bundle by checking the visited urls for the
+        // bundle url itself.
+        visitedUrls.has(importBundle.url) &&
```
consider making a well-named variable to hold this boolean value to summarize what it indicates. I like the comments :)

usergenic replied:
Ok.  Did it!

justinfagnani commented on src/test/deps-index_test.ts:
```diff
@@ -23,66 +23,61 @@ import {buildDepsIndex} from '../deps-index';
+      const expectedEntrypointsToDeps = new Map([
+        [entrypoint, new Set([entrypoint])],
+        [lazyImport1, new Set([lazyImport1, shared1, shared2])],
+        [lazyImport2, new Set([lazyImport2, shared1, shared2, shared3])],
```
I'd like to see a test with a lazy import that itself has a lazy import that has dependencies, so test whether or not lazy imports need to be fully processed as entrypoints.

usergenic replied:
I know we said in-person that we didn't need this but I added it anyways.  Test upgraded with lazily imported imports of lazy imports.

usergenic commented:
Okay @justinfagnani lazy imports PR feedback has been addressed.  PTAL!
